File: src/examples/python/laguerre.py
# add the path to the pyavro module using sys
import random
import sys
sys.path.insert(0,'../../../build/release/lib/')

# import pyavro
import pyavro

number = 4
dim    = number
udim   = dim -1
ctx    = pyavro.Context(number,dim,udim)

nb_points = 1 * 10**3
x = [ random.uniform(0,1) for i in range(dim*nb_points) ]
w = [ 0.0 for i in range(nb_points) ]

x = ctx.compute_laguerre( x , w , 10 )
unit_mass = 1./nb_points
mass = [ unit_mass for i in range(nb_points) ]

'''
total_mass = sum(mass)
for k in range(nb_points):
    nb_left = nb_points -k
    dm = random.uniform( 0.001 ,total_mass/nb_left)
    mass[k] = dm
    total_mass -= mass[k]
'''

w = ctx.compute_optimal_transport( x , mass , w , 10 )

vertices,polytopes,nv_per_elem = pyavro.retrieve_polytopes(ctx)

nb_vertices  = len(vertices)/dim
nb_polytopes = len(nv_per_elem)

'''
i = 0
for k in range(nb_polytopes):
    print('polytope[',k,']: (',end='')
    for j in range(nv_per_elem[k]):
        print(polytopes[i],end=' ')
        i += 1
    print(')')
'''

ctx.plot()

File: src/examples/laguerre.cpp
#include <avro.h>

#include <math.h>
#include <cstdlib>
#include <assert.h>

int
main( int argc , char** argv )
{
  using avro::index_t;
  using avro::coord_t;
  using avro::real_t;

  // setup the context
  coord_t number = 3;
  coord_t dim = number;
  coord_t udim = dim-1;
  avro::Context context(number,dim,udim);

  index_t nb_points = 100;

  std::vector<real_t> x( nb_points*dim );
  std::vector<real_t> w( nb_points , 0.0 );

  for (index_t k = 0; k < nb_points; k++)
  for (coord_t d = 0; d < dim; d++)
    x[dim*k+d] = real_t(rand())/real_t(RAND_MAX);

  context.compute_laguerre( x , w , 0 );

  std::vector<real_t> vertices;
  std::vector<index_t> polytopes;
  std::vector<index_t> nv_per_elem;
  context.retrieve_polytopes( vertices , polytopes , nv_per_elem );

  index_t nb_vertices = vertices.size()/dim;
  for (index_t k = 0; k < nb_vertices; k++) {
    printf("voronoi vertex[%lu] = ( ",k);
    for (coord_t d = 0; d < dim; d++)
      printf("%g ",vertices[k*dim+d]);
    printf(")\n");
  }

  index_t nb_polytopes = nv_per_elem.size();
  index_t i = 0;
  for (index_t k = 0; k < nb_polytopes; k++) {
    printf("polytope[%lu] = ( ",k);
    for (index_t j = 0; j < nv_per_elem[k]; j++)
      printf("%lu ",polytopes[i++]);
    printf(")\n");
  }

  assert( nb_polytopes == nb_points );

  return 0;
}
